File: tests/run_bsroformer_models.py
import argparse
import json
import os
import time

import torch

from workers.bsroformer_worker import load_model_impl, separate_impl, load_models_registry


def parse_args():
    parser = argparse.ArgumentParser(description="Run BS-RoFormer models on one input file")
    parser.add_argument("--input", required=True, help="Input audio file path")
    parser.add_argument("--models-dir", required=True, help="Models directory (contains models.json)")
    parser.add_argument("--output-dir", required=True, help="Output directory for results")
    parser.add_argument("--overlap", type=int, default=2, help="Overlap factor")
    parser.add_argument("--batch-size", type=int, default=2, help="Batch size")
    parser.add_argument("--use-tta", action="store_true", help="Enable TTA")
    parser.add_argument("--fast", action="store_true", help="Use vectorized chunking")
    parser.add_argument("--limit", type=int, default=0, help="Limit number of models (0 = all)")
    return parser.parse_args()


def main():
    args = parse_args()
    if not os.path.exists(args.input):
        raise FileNotFoundError(f"Input not found: {args.input}")

    os.makedirs(args.output_dir, exist_ok=True)
    report_path = os.path.join(args.output_dir, "report.json")

    device = "cuda" if torch.cuda.is_available() else "cpu"
    models = load_models_registry(args.models_dir)

    results = []
    model_names = list(models.keys())
    if args.limit and args.limit > 0:
        model_names = model_names[:args.limit]

    for name in model_names:
        start = time.time()
        status = "ok"
        error = None
        try:
            model, config, model_info = load_model_impl(name, args.models_dir)
            separate_impl(
                model,
                config,
                args.input,
                os.path.join(args.output_dir, name),
                model_info,
                device=device,
                overlap=args.overlap,
                batch_size=args.batch_size,
                use_tta=args.use_tta,
                output_format="wav",
                pcm_type="FLOAT",
                extract_instrumental=False,
                selected_stems=None,
                two_stems=None,
                use_fast=args.fast,
            )
        except Exception as exc:
            status = "error"
            error = str(exc)
        elapsed = round(time.time() - start, 2)
        results.append(
            {
                "model": name,
                "status": status,
                "elapsed_s": elapsed,
                "error": error,
            }
        )

        with open(report_path, "w", encoding="utf-8") as f:
            json.dump(results, f, indent=2)

    print(f"Report written to: {report_path}")


if __name__ == "__main__":
    main()

File: tests/run_bsroformer_models.js
const fs = require("fs");
const path = require("path");
const { spawnSync } = require("child_process");

function parseArgs() {
  const args = process.argv.slice(2);
  const out = {};
  for (let i = 0; i < args.length; i += 1) {
    const key = args[i];
    if (!key.startsWith("--")) continue;
    const value = args[i + 1];
    if (value && !value.startsWith("--")) {
      out[key.slice(2)] = value;
      i += 1;
    } else {
      out[key.slice(2)] = true;
    }
  }
  return out;
}

function ensureDir(dir) {
  fs.mkdirSync(dir, { recursive: true });
}

function main() {
  const args = parseArgs();
  const input = args.input;
  const modelsDir = args["models-dir"];
  const outputDir = args["output-dir"];
  const pythonExe = args.python;
  const limit = args.limit ? parseInt(args.limit, 10) : 0;

  if (!input || !modelsDir || !outputDir || !pythonExe) {
    console.error("Usage: node run_bsroformer_models.js --input <wav> --models-dir <dir> --output-dir <dir> --python <python.exe> [--limit N] [--overlap N] [--batch-size N] [--use-tta] [--fast]");
    process.exit(2);
  }

  const registryPath = path.join(modelsDir, "models.json");
  if (!fs.existsSync(registryPath)) {
    console.error(`models.json not found: ${registryPath}`);
    process.exit(2);
  }

  const models = JSON.parse(fs.readFileSync(registryPath, "utf-8"));
  const modelNames = Object.keys(models);
  const selected = limit > 0 ? modelNames.slice(0, limit) : modelNames;

  ensureDir(outputDir);
  const reportPath = path.join(outputDir, "report.json");
  const results = [];

  for (const modelName of selected) {
    const modelOutDir = path.join(outputDir, modelName);
    ensureDir(modelOutDir);

    const cmdArgs = [
      path.join(__dirname, "..", "workers", "bsroformer_worker.py"),
      "--model",
      modelName,
      "--models-dir",
      modelsDir,
      "--input_folder",
      input,
      "--store_dir",
      modelOutDir,
    ];

    if (args.overlap) {
      cmdArgs.push("--overlap", String(args.overlap));
    }
    if (args["batch-size"]) {
      cmdArgs.push("--batch-size", String(args["batch-size"]));
    }
    if (args["use-tta"]) {
      cmdArgs.push("--use_tta");
    }
    if (args.fast) {
      cmdArgs.push("--fast");
    }

    const start = Date.now();
    const result = spawnSync(pythonExe, cmdArgs, { stdio: "inherit" });
    const elapsed = Math.round((Date.now() - start) / 1000);

    results.push({
      model: modelName,
      status: result.status === 0 ? "ok" : "error",
      elapsed_s: elapsed,
      exit_code: result.status,
    });

    fs.writeFileSync(reportPath, JSON.stringify(results, null, 2));
  }

  console.log(`Report written to: ${reportPath}`);
}

main();
